### DISCUSSION TITLE
Memory usage

### DISCUSSION THREAD

**Discussion by MorsMortium** [ImageMagick/ImageMagick#5925] [Help] [answered]
### ImageMagick version

7.1.0.55-2

### Operating system, version and so on

Manjaro linux

### Description

Hi. I'm rewriting a program of mine from python to c. It is done file by file, with python c extensions, so this function, I'll show, is in such an environment.

```
bool ImageConvert(char *FilePath)
{
	// This function is responsible for converting a file into GTK compatible
	// type and returning if it was able to do it

	// Check for shutdown, enter lock
	if (ImageIncrease())
		return false;

	// Create wand, exit if unsuccessful
	MagickWand *Wand = NewMagickWand();
	if (Wand == NULL) {
		ImageDecrease();
		return false;
	}
	// Add .gif to name
	char LocalPath[strlen(FilePath) + 5];
	sprintf(LocalPath, "%s.gif", FilePath);

	// Check if gif version exists
	MagickBooleanType Status = MagickPingImage(Wand, LocalPath);

	// If not, load regular version
	if (Status != MagickTrue)
		Status = MagickPingImage(Wand, FilePath);

	if (Status == MagickTrue) {
		// Get format, convert to lowercase
		char *Format = MagickGetImageFormat(Wand);
		for (int Index = 0; Index < (int) strlen(Format); ++Index)
			Format[Index] = tolower(Format[Index]);

		// Go through every format signify and exit if file format is good
		bool Good = false;
		int Index = 0;
		while (ImageTypes[Index] != NULL) {
			if (strcmp(ImageTypes[Index], Format) == 0) {
				Good = true;
				break;
			}
			++Index;
		}

		free(Format);
		if (!Good) {

			// Empty wand, load full image
			ClearMagickWand(Wand);
			Status = MagickReadImage(Wand, FilePath);
			if (Status == MagickTrue) {

				// If image is a single frame
				if (MagickGetNumberImages(Wand) == 1) {

					// Save as png
					MagickSetFormat(Wand, "PNG");
					MagickWriteImage(Wand, FilePath);
				} else {

					// Save as gif under new name
					MagickSetFormat(Wand, "GIF");
					MagickWriteImages(Wand, LocalPath, true);
				}
			}
		}
	}
	// Destroy wand, leave lock, return success or failure
	DestroyMagickWand(Wand);
	ImageDecrease();
	return Status == MagickTrue;
}
```
MagickWandGenesis and MagickWandTerminus are called in different functions, on program startup and shutdown.

My problem with this is that after DestroyMagickWand it still uses huge amounts of memory. The python version, using PIL had a similar problem, there I solved it with offloading everything to a process. With the process, the whole app used 95 mb of ram, for a certain amount of images (haven't counted, but it was the exact same set of pictures, with the same number converted), while with IM, it was 270 mb. 

Before I discovered MagickPingImage, and used MagickReadImage, even for format detecting, it used even more ram, and kept it after DestroyMagickWand. 

Until now I tried restarting the environment with MagickWandGenesis and MagickWandTerminus per 10 pictures, rewrote the code to use magickcore instead, tried python garbage collection (though c extensions manage their own memory), tried ClearMagickWand before DestroyMagickWand.

Any tips on how I could make it release memory would be appreciated.

**Comment by urban-warrior** [comment, marked as answer]
Freeing memory does not always return the memory to the system.  Use `valgrind` to verify no memory leaks in your code. 
You can minimize the memory footprint by pushing the image pixels to disk rather than memory with SetMagickResourceLimit().  Set the memory and map limits to something reasonable, perhaps 2GB and 4GB respectively.  Note, caching pixels to disk reduces the memory footprint, but is slower.  See https://imagemagick.org/script/resources.php and https://imagemagick.org/script/architecture.php for details about  ImageMagick resource usage and limits.

**Reply by MorsMortium** [reply]
In c I'm pretty sure it does. And even if not, I haven't seen a program, that stored gbs of ram, that it free'd already. Valgrind only talks about some bytes, in different places, but i traced the hundreds of mb to this function, as I mentioned replacing MagickReadImage with MagickPingImage didn't read the files and then keep the memory. So what I get from this, is there no way to force IM, to release memory it no longer uses?

**Reply by urban-warrior** [reply]
Perhaps there is a leak in your codeset.  Valgrind should be able to identify it if there is.  We run `valgrind` against ImageMagick on a regular basis to help ensure there are no memory leaks and we use standard system calls to allocate memory.  Some memory algorithms return memory to the system, some do not.  We would need evidence that ImageMagick memory is leaking (e.g. a `valgrind` report) before we can assist further.  Here is a sample run that shows no memory leaks with `valgrind`:

```
$ valgrind magick rose: -resize 512x512 rose.pnm
==2391400== Memcheck, a memory error detector
==2391400== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==2391400== Using Valgrind-3.20.0 and LibVEX; rerun with -h for copyright info
==2391400== Command: magick rose: -resize 512x512 rose.pnm
==2391400== 
==2391400== 
==2391400== HEAP SUMMARY:
==2391400==     in use at exit: 23,236 bytes in 28 blocks
==2391400==   total heap usage: 1,533 allocs, 1,505 frees, 3,684,868 bytes allocated
==2391400== 
==2391400== LEAK SUMMARY:
==2391400==    definitely lost: 0 bytes in 0 blocks
==2391400==    indirectly lost: 0 bytes in 0 blocks
==2391400==      possibly lost: 1,280 bytes in 4 blocks
==2391400==    still reachable: 21,956 bytes in 24 blocks
==2391400==         suppressed: 0 bytes in 0 blocks
==2391400== Rerun with --leak-check=full to see details of leaked memory
==2391400== 
==2391400== For lists of detected and suppressed errors, rerun with: -s
==2391400== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Notice the `definitely lost` of 0 byte, suggesting no leaks.  Run a similar check against your codeset and identify any leaks.  If so, make sure its not a flow issue with your code.

**Reply by MorsMortium** [reply]
Okay, after extensive testing (writing a small c extension library, that loads and converts a single image, and then trying that from both multithreaded c and python code, with and without valgrind) it turned out that it's both python not releasing memory and c in general (python almost always kept more memory, when finishing, but the pure c function also didn't release all). And it was also confirmed, that the memory would be released, because when coverting the same number of images, single threaded, it didn't keep more memory, than needed for one image. So with all this, I opted for also using a process in the c version, to don't keep gbs of memory around.
